service/dap/server.go

```diff
@@ -451,6 +451,50 @@ func (s *Server) onContinueRequest(request *dap.ContinueRequest) {
 	s.doContinue()
 }
 
+func (s *Server) onThreadsRequest(request *dap.ThreadsRequest) {
+	if s.debugger == nil {
+		s.sendErrorResponse(request.Request, UnableToDisplayThreads, "Unable to display threads", "debugger is nil")
+		return
+	}
+	gs, _, err := s.debugger.Goroutines(0, 0)
```

Comment:
This is an interesting point. I will reach out to my contacts for additional input. They do make a lot of these requests and the current adaptor translates them into the get-all delve rpc call. And what is even more unfortunate is that I see many back-to-back dups in the logs with the exact same responses. Is your concern that the debugger will be busy dealing with each one and unable to respond to other requests or that sending the long list over the connection would add too much latency?

Reply:
Both. Also that the UI will be inefficient in handling a long list of goroutines and add further latency.

Reply:
The [overview documentation ](https://microsoft.github.io/debug-adapter-protocol/overview), "Whenever the generic debugger receives a stopped or a thread event, the development tool requests all threads that exist at that point in time. Thread events are optional, but a debug adapter can send them to force the development tool to update the threads UI dynamically even when not in a stopped state. If a debug adapter decides not to emit Thread events, the thread UI in the development tool will only update if a stopped event is received."

Do you therefore recommend that I not implement the threads events, not even in the case where the program does not stop on entry, so no threads are requested and displayed then until another stop is reached?

Reply:
Also, if you were to design a custom API for go for this, what would you do? Would you only allow calls upon a user request instead of automatically on any stop? Or try to differentiate between user-defined goroutines, runtime goroutines and any hidden library goroutines and limit the list by default to only the ones the user defined and hence cares about?

Reply:
> Do you therefore recommend that I not implement the threads events, not even in the case where the program does not stop on entry, so no threads are requested and displayed then until another stop is reached?

I don't think there's any way to implement the thread event.

> Also, if you were to design a custom API for go for this, what would you do?

I think it should be more like the stack trace API where the client can request a limited number of threads and then there's a mechanism to request more.

Reply:
vscode-go currently implements thread events when stopped, which is why I had them in mind next. It appears that there was a time when threads request was not used to update the UI, so users used additional thread events even in stopped state (see  [this issue](https://github.com/Microsoft/vscode/issues/40) and [this issue](https://github.com/microsoft/vscode/issues/19858)). But I do not think this is no longer necessary. I am scrapping my delve [PR]( https://github.com/polinasok/delve/pull/7) for thread events and instead will change vscode-go to no longer issue them either. 

Do I understand correctly that ListGoroutines({count: 1}) will return just the current goroutine?

Reply:
> Do I understand correctly that ListGoroutines({count: 1}) will return the current goroutine or is it always goroutine 1?

Always 1 but even that is just a consequence of how the go runtime works. The current goroutine is in DebuggerState.SelectedGoroutine (or debugger.target.SelectedGoroutine()). Note that it can be nil (at entry and when the current thread is executing non-go code).

Reply:
Looking at that code, what they are doing is picking the first goroutine if DebuggerState.SelectedGoroutine is nil. I think in that case a better choice would be to use GoroutineID == 0.

Reply:
This is the vscode-go PR: https://github.com/microsoft/vscode-go/pull/3145

Reply:
Also, how can I trigger this corner case where there is no selected goroutine in debug state?

Reply:
Launch a program, do not execute `Continue`, the initial state of delve will be with SelectedGoroutine == nil.

Reply:
Ah, I see. Is this the only case? Then it is the same case where I am forced to send back a dummy thread when stopping on entry, isn't it? So if the debug state has nil, won't the list of goroutines be empty as well? They use the selected goroutine id to decide which goroutine to highlight as stopped on in the UI by sticking it into the stopped event. It doesn't seem that id 0 would be of much help. What does that represent anyway? I am still not quite clear about your earlier suggestion.

Reply:
> Ah, I see. Is this the only case? Then it is the same case where I am forced to send back a dummy thread when stopping on entry, isn't it? So if the debug state has nil, won't the list of goroutines be empty as well?

No, this will happen anytime we can't read the current goroutine from TLS. Another example is if cgo starts a thread that only executed C code and that thread gets selected as the current thread (either because of a breakpoint, a signal or because of a manual stop).

> They use the selected goroutine id to decide which goroutine to highlight as stopped on in the UI by sticking it into the stopped event. It doesn't seem that id 0 would be of much help. What does that represent anyway? I am still not quite clear about your earlier suggestion.

It won't help you with that, it would help you if you used it to request local variables, stack traces or evaluate expressions.

Reply:
The id in the stopped event not only helps the UI mark which thread is paused, but is also used in subsequent stack trace and variable requests to reflect that state of things in the UI as well. 

What exactly does goroutine id 0 represent? My understanding was that goroutines are numbered from 1 up.

I see no currentGoroutine set when I use pause (which results in halt rpc). In that case, the code gets a list of goroutines and uses goroutine[0], which as far as I can tell is always id=1. The code expects the entire goroutine record, but I hacked it real fast and just put 0 for id with the rest of the fields coming from goroutine[0]. Is that what you meant? That doesn't result in anything useful. No stacktraces or variables are requested/displayed with id 0.

Reply:
Goroutine id 0 is used for a series of special things in the go runtime, we're also treating it specially by allowing it to represent the current thread when the current thread doesn't have an associated goroutine.

>The code expects the entire goroutine record, but I hacked it real fast and just put 0 for id with the rest of the fields coming from goroutine[0]. Is that what you meant?

The proper way to do it would be to leave most fields empty and copy the current location from the CurrentThread field.

> That doesn't result in anything useful. No stacktraces or variables are requested/displayed with id 0.

This could be a bug in delve but it's hard to say because what happens when a pause is requested is somewhat random and depends on the program being run as well as the operating system.

Reply:
> Do you have any contacts on the VSCode side of things? This API is very unfortunate, it works ok for actual threads, which are bound to be few in number, but there could be _a lot_ of goroutines and requesting all of them after every user action will slow down debugging a lot.

Please see https://github.com/microsoft/debug-adapter-protocol/issues/159.

Reply:
Thanks.